[FLINK-25035][runtime] Shuffle service supports consuming subpartition range

wanglijie95 opened the pull request:
## What is the purpose of the change
Let shuffle service supports consuming subpartition range. In adaptive batch scheduler, the shuffle service needs to support a SingleInputGate to consume  a certain range of subpartitions.

## Brief change log
68e3afaf272233bd0ca180eba092d06325fd2533 Move consumedSubpartitionIndex from SingleInputGate to InputChannel
41eeab31ee52a29d71acf018f9d1ef5a7750bb85 SingleInputGate supports consuming subpartition range

## Verifying this change
This change is already covered by existing tests.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (**no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (**no**)
  - The serializers: (**no**)
  - The runtime per-record code paths (performance sensitive): (**no**)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (**no**)
  - The S3 file system connector: (**no**)

## Documentation

  - Does this pull request introduce a new feature? (**no**)
  - If yes, how is the feature documented? (**not applicable**)

zhuzhurk reviewed (commented):
Thanks for opening this PR! @wanglijie95 
The changes of commit 12880feeedbbba9d3beb93225843218f4d6776f6 looks generally good to me. I only have 2 minor comments and please take a look.

wanglijie95 commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/InputChannel.java:
```diff
@@ -54,6 +54,8 @@
 
     protected final ResultPartitionID partitionId;
 
+    protected final int subpartitionIdx;
```
fixed

wanglijie95 commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/SingleInputGateFactory.java:
```diff
@@ -173,20 +177,35 @@ private void createInputChannels(
         ShuffleDescriptor[] shuffleDescriptors =
                 inputGateDeploymentDescriptor.getShuffleDescriptors();
 
         // Create the input channels. There is one input channel for each consumed partition.
```
fixed

wanglijie95 commented:
> Thanks for opening this PR! @wanglijie95 The changes of commit [12880fe](https://github.com/apache/flink/commit/12880feeedbbba9d3beb93225843218f4d6776f6) looks generally good to me. I only have 2 minor comments and please take a look.

Thanks for review @zhuzhurk .  I've addressed all your comments and split the changes into 2 commits, please review again.

wanglijie95 commented:
@wsry This change involves some modifications to the network. Would you review it?

zhuzhurk commented on flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/consumer/SingleInputGateTest.java:
```diff
@@ -645,13 +616,13 @@ public void testRequestBackoffConfiguration() throws Exception {
                                 }
                             });
-            InputChannel localChannel = channelMap.get(partitionIds[0]);
+            InputChannel localChannel = channelMap.get(new SubpartitionInfo(partitionIds[0], 0));
```
It's better to create a testing method `createSubpartitionInfo(partitionId)`.

wanglijie95 replied:
fixed

wanglijie95 commented:
@flinkbot run azure

wsry commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/InputChannel.java:
```diff
@@ -94,6 +97,7 @@ protected InputChannel(
         this.inputGate = checkNotNull(inputGate);
         this.channelInfo = new InputChannelInfo(inputGate.getGateIndex(), channelIndex);
         this.partitionId = checkNotNull(partitionId);
+        this.consumedSubpartitionIndex = consumedSubpartitionIndex;
```
nit: maybe add an ```>=0``` argument check

wanglijie95 replied:
Done

wsry commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/InputChannel.java:
```diff
@@ -168,8 +176,7 @@ protected void notifyBufferAvailable(int numAvailableBuffers) throws IOException
      * <p>The queue index to request depends on which sub task the channel belongs to and is
      * specified by the consumer of this channel.
```
```and is specified by the consumer of this channel.``` This java doc seems out of date. Maybe delete this java doc and add the java doc to the ```consumedSubpartitionIndex``` filed.

wanglijie95 replied:
Done

wsry commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/SingleInputGate.java:
```diff
@@ -1073,7 +1085,32 @@ private boolean queueChannelUnsafe(InputChannel channel, boolean priority) {
+        }
+
+        @Override
+        public boolean equals(Object obj) {
```
nit: maybe return true directly if this == obj.

wanglijie95 replied:
Done

wsry commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/SingleInputGateFactory.java:
```diff
@@ -224,15 +245,25 @@ private InputChannel createInputChannel(
 
+    private static int calculateNumChannels(
+            int numShuffleDescriptors, SubpartitionIndexRange subpartitionIndexRange) {
+        return numShuffleDescriptors
```
Do we need a overflow check?

wanglijie95 replied:
Added.

wsry commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/SingleInputGate.java:
```diff
@@ -1073,7 +1085,32 @@ private boolean queueChannelUnsafe(InputChannel channel, boolean priority) {
+        private final int subpartitionIndex;
+
+        SubpartitionInfo(IntermediateResultPartitionID partitionID, int subpartitionIndex) {
+            this.partitionID = partitionID;
```
nit: maybe add a not null check

wanglijie95 replied:
Done

wsry commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/SingleInputGate.java:
```diff
@@ -1073,7 +1085,32 @@ private boolean queueChannelUnsafe(InputChannel channel, boolean priority) {
+
+        SubpartitionInfo(IntermediateResultPartitionID partitionID, int subpartitionIndex) {
+            this.partitionID = partitionID;
+            this.subpartitionIndex = subpartitionIndex;
```
nit: maybe add an >=0 argument check

wanglijie95 replied:
Done.

wsry commented on flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/consumer/SingleInputGateTest.java:
```diff
@@ -630,9 +601,9 @@ public void testRequestBackoffConfiguration() throws Exception {
             closer.register(gate::close);
 
-            assertEquals(gateDesc.getConsumedPartitionType(), gate.getConsumedPartitionType());
+            assertEquals(ResultPartitionType.PIPELINED, gate.getConsumedPartitionType());
```
Is it a bugfix?

wsry commented on flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/consumer/SingleInputGateTest.java:
```diff
@@ -834,14 +817,82 @@ public void testUpdateUnknownInputChannel() throws Exception {
     }
 
+    @Test
+    public void testUpdateUnknownChannelWithSubpartitionIndexRange()
```
Maybe it's better to also cover the partition request with subpartition range scenario.

wanglijie95 replied:
Done

wsry commented:
@wanglijie95 Thanks for the contribution. The code looks generally good to me and I only left several tiny comments.

wanglijie95 commented:
> @wanglijie95 Thanks for the contribution. The code looks generally good to me and I only left several tiny comments.

Thanks for review @wsry  . I've addressed all your comments. Please help to review again.

wanglijie95 commented:
Squashed commits.

zhuzhurk reviewed (approved):
LGTM.

wsry commented:
@wanglijie95 Thanks for the update. Changes LGTM.

zhuzhurk commented:
Merging.
